### ISSUE TITLE
Content negotiation in Plone via dynamic layers

### ISSUE THREAD

**Issue by metatoaster** [plone/plone.restapi#19]
Are there any existing plone.\* package that can provide content negotiation?  This is needed to provide web-services with the constraints set forth by HATEOAS (which I noted that it is finally included as a consideration, which satisfied some of my original concerns about Plone's approach of web-services).

One way to get around this issue specific to Plone in one of my projects is to have a subscriber to the `IBeforeTraverseEvent` event for the site root. The subscriber will then dynamically mark the request with layer(s) from matching rule(s) (i.e. the mime-type at hand) as set forth by any of the registered layer utilities.  These layer utilities can then be provided by the implementation packages for any of the JSON web-services oriented content types to register their availability within Plone.

Anyway, this issue is mostly a question on what to do, as I noticed that the design documentation haven't really addressed this point yet or any possible approaches to solve this problem, or whether this is already the approach already being considered on some other discussion lists elsewhere.

**Comment by tisto** [comment]
Content negotiation is on our todo list. Though, we plan to focus on json exclusively first. I guess it should be possible to amend our IBeforeTraverseEvent hook to do pluggable content negotiation:

https://github.com/plone/plone.rest/blob/master/src/plone/rest/events.py

Ideas, requirements, code examples or any other contributions are more than welcome!

**Comment by metatoaster** [comment]
https://github.com/PMR2/pmr2.layer/tree/0.1/pmr2/layer

This is what I did some time ago but it does not address the complete syntax as documented in the HTTP/1.1 Header Field Definition ([Section 14.1](http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html)), but in general this shouldn't be a deal breaker as clients generally only want to deal with one specific format.  For the most part, I did this mostly to provide canonical unique URL per resource (regardless of method of access) and versioning of the API (version 0 did not use any standard, version 1 moved onto using Collections+JSON in our case, however I didn't know better and didn't use `;version=0` to denote version info in my case).

In any case, the entire project plus the test case is basically a set of code examples that can be adapted by Plone to make use of.  For the most part this seems to work well enough for the project using that.
